<!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <title>Metabarcoding workshop</title>
        <style>
/* From extension vscode.github */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.vscode-dark img[src$=\#gh-light-mode-only],
.vscode-light img[src$=\#gh-dark-mode-only] {
	display: none;
}

/* From extension vscode.markdown-math */
@font-face{font-family:KaTeX_AMS;font-style:normal;font-weight:400;src:url(fonts/KaTeX_AMS-Regular.woff2) format("woff2"),url(fonts/KaTeX_AMS-Regular.woff) format("woff"),url(fonts/KaTeX_AMS-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Caligraphic;font-style:normal;font-weight:700;src:url(fonts/KaTeX_Caligraphic-Bold.woff2) format("woff2"),url(fonts/KaTeX_Caligraphic-Bold.woff) format("woff"),url(fonts/KaTeX_Caligraphic-Bold.ttf) format("truetype")}@font-face{font-family:KaTeX_Caligraphic;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Caligraphic-Regular.woff2) format("woff2"),url(fonts/KaTeX_Caligraphic-Regular.woff) format("woff"),url(fonts/KaTeX_Caligraphic-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Fraktur;font-style:normal;font-weight:700;src:url(fonts/KaTeX_Fraktur-Bold.woff2) format("woff2"),url(fonts/KaTeX_Fraktur-Bold.woff) format("woff"),url(fonts/KaTeX_Fraktur-Bold.ttf) format("truetype")}@font-face{font-family:KaTeX_Fraktur;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Fraktur-Regular.woff2) format("woff2"),url(fonts/KaTeX_Fraktur-Regular.woff) format("woff"),url(fonts/KaTeX_Fraktur-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Main;font-style:normal;font-weight:700;src:url(fonts/KaTeX_Main-Bold.woff2) format("woff2"),url(fonts/KaTeX_Main-Bold.woff) format("woff"),url(fonts/KaTeX_Main-Bold.ttf) format("truetype")}@font-face{font-family:KaTeX_Main;font-style:italic;font-weight:700;src:url(fonts/KaTeX_Main-BoldItalic.woff2) format("woff2"),url(fonts/KaTeX_Main-BoldItalic.woff) format("woff"),url(fonts/KaTeX_Main-BoldItalic.ttf) format("truetype")}@font-face{font-family:KaTeX_Main;font-style:italic;font-weight:400;src:url(fonts/KaTeX_Main-Italic.woff2) format("woff2"),url(fonts/KaTeX_Main-Italic.woff) format("woff"),url(fonts/KaTeX_Main-Italic.ttf) format("truetype")}@font-face{font-family:KaTeX_Main;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Main-Regular.woff2) format("woff2"),url(fonts/KaTeX_Main-Regular.woff) format("woff"),url(fonts/KaTeX_Main-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Math;font-style:italic;font-weight:700;src:url(fonts/KaTeX_Math-BoldItalic.woff2) format("woff2"),url(fonts/KaTeX_Math-BoldItalic.woff) format("woff"),url(fonts/KaTeX_Math-BoldItalic.ttf) format("truetype")}@font-face{font-family:KaTeX_Math;font-style:italic;font-weight:400;src:url(fonts/KaTeX_Math-Italic.woff2) format("woff2"),url(fonts/KaTeX_Math-Italic.woff) format("woff"),url(fonts/KaTeX_Math-Italic.ttf) format("truetype")}@font-face{font-family:"KaTeX_SansSerif";font-style:normal;font-weight:700;src:url(fonts/KaTeX_SansSerif-Bold.woff2) format("woff2"),url(fonts/KaTeX_SansSerif-Bold.woff) format("woff"),url(fonts/KaTeX_SansSerif-Bold.ttf) format("truetype")}@font-face{font-family:"KaTeX_SansSerif";font-style:italic;font-weight:400;src:url(fonts/KaTeX_SansSerif-Italic.woff2) format("woff2"),url(fonts/KaTeX_SansSerif-Italic.woff) format("woff"),url(fonts/KaTeX_SansSerif-Italic.ttf) format("truetype")}@font-face{font-family:"KaTeX_SansSerif";font-style:normal;font-weight:400;src:url(fonts/KaTeX_SansSerif-Regular.woff2) format("woff2"),url(fonts/KaTeX_SansSerif-Regular.woff) format("woff"),url(fonts/KaTeX_SansSerif-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Script;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Script-Regular.woff2) format("woff2"),url(fonts/KaTeX_Script-Regular.woff) format("woff"),url(fonts/KaTeX_Script-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Size1;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Size1-Regular.woff2) format("woff2"),url(fonts/KaTeX_Size1-Regular.woff) format("woff"),url(fonts/KaTeX_Size1-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Size2;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Size2-Regular.woff2) format("woff2"),url(fonts/KaTeX_Size2-Regular.woff) format("woff"),url(fonts/KaTeX_Size2-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Size3;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Size3-Regular.woff2) format("woff2"),url(fonts/KaTeX_Size3-Regular.woff) format("woff"),url(fonts/KaTeX_Size3-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Size4;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Size4-Regular.woff2) format("woff2"),url(fonts/KaTeX_Size4-Regular.woff) format("woff"),url(fonts/KaTeX_Size4-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Typewriter;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Typewriter-Regular.woff2) format("woff2"),url(fonts/KaTeX_Typewriter-Regular.woff) format("woff"),url(fonts/KaTeX_Typewriter-Regular.ttf) format("truetype")}.katex{text-rendering:auto;font:normal 1.21em KaTeX_Main,Times New Roman,serif;line-height:1.2;text-indent:0}.katex *{-ms-high-contrast-adjust:none!important;border-color:currentColor}.katex .katex-version:after{content:"0.16.2"}.katex .katex-mathml{clip:rect(1px,1px,1px,1px);border:0;height:1px;overflow:hidden;padding:0;position:absolute;width:1px}.katex .katex-html>.newline{display:block}.katex .base{position:relative;white-space:nowrap;width:-webkit-min-content;width:-moz-min-content;width:min-content}.katex .base,.katex .strut{display:inline-block}.katex .textbf{font-weight:700}.katex .textit{font-style:italic}.katex .textrm{font-family:KaTeX_Main}.katex .textsf{font-family:KaTeX_SansSerif}.katex .texttt{font-family:KaTeX_Typewriter}.katex .mathnormal{font-family:KaTeX_Math;font-style:italic}.katex .mathit{font-family:KaTeX_Main;font-style:italic}.katex .mathrm{font-style:normal}.katex .mathbf{font-family:KaTeX_Main;font-weight:700}.katex .boldsymbol{font-family:KaTeX_Math;font-style:italic;font-weight:700}.katex .amsrm,.katex .mathbb,.katex .textbb{font-family:KaTeX_AMS}.katex .mathcal{font-family:KaTeX_Caligraphic}.katex .mathfrak,.katex .textfrak{font-family:KaTeX_Fraktur}.katex .mathtt{font-family:KaTeX_Typewriter}.katex .mathscr,.katex .textscr{font-family:KaTeX_Script}.katex .mathsf,.katex .textsf{font-family:KaTeX_SansSerif}.katex .mathboldsf,.katex .textboldsf{font-family:KaTeX_SansSerif;font-weight:700}.katex .mathitsf,.katex .textitsf{font-family:KaTeX_SansSerif;font-style:italic}.katex .mainrm{font-family:KaTeX_Main;font-style:normal}.katex .vlist-t{border-collapse:collapse;display:inline-table;table-layout:fixed}.katex .vlist-r{display:table-row}.katex .vlist{display:table-cell;position:relative;vertical-align:bottom}.katex .vlist>span{display:block;height:0;position:relative}.katex .vlist>span>span{display:inline-block}.katex .vlist>span>.pstrut{overflow:hidden;width:0}.katex .vlist-t2{margin-right:-2px}.katex .vlist-s{display:table-cell;font-size:1px;min-width:2px;vertical-align:bottom;width:2px}.katex .vbox{align-items:baseline;display:inline-flex;flex-direction:column}.katex .hbox{width:100%}.katex .hbox,.katex .thinbox{display:inline-flex;flex-direction:row}.katex .thinbox{max-width:0;width:0}.katex .msupsub{text-align:left}.katex .mfrac>span>span{text-align:center}.katex .mfrac .frac-line{border-bottom-style:solid;display:inline-block;width:100%}.katex .hdashline,.katex .hline,.katex .mfrac .frac-line,.katex .overline .overline-line,.katex .rule,.katex .underline .underline-line{min-height:1px}.katex .mspace{display:inline-block}.katex .clap,.katex .llap,.katex .rlap{position:relative;width:0}.katex .clap>.inner,.katex .llap>.inner,.katex .rlap>.inner{position:absolute}.katex .clap>.fix,.katex .llap>.fix,.katex .rlap>.fix{display:inline-block}.katex .llap>.inner{right:0}.katex .clap>.inner,.katex .rlap>.inner{left:0}.katex .clap>.inner>span{margin-left:-50%;margin-right:50%}.katex .rule{border:0 solid;display:inline-block;position:relative}.katex .hline,.katex .overline .overline-line,.katex .underline .underline-line{border-bottom-style:solid;display:inline-block;width:100%}.katex .hdashline{border-bottom-style:dashed;display:inline-block;width:100%}.katex .sqrt>.root{margin-left:.27777778em;margin-right:-.55555556em}.katex .fontsize-ensurer.reset-size1.size1,.katex .sizing.reset-size1.size1{font-size:1em}.katex .fontsize-ensurer.reset-size1.size2,.katex .sizing.reset-size1.size2{font-size:1.2em}.katex .fontsize-ensurer.reset-size1.size3,.katex .sizing.reset-size1.size3{font-size:1.4em}.katex .fontsize-ensurer.reset-size1.size4,.katex .sizing.reset-size1.size4{font-size:1.6em}.katex .fontsize-ensurer.reset-size1.size5,.katex .sizing.reset-size1.size5{font-size:1.8em}.katex .fontsize-ensurer.reset-size1.size6,.katex .sizing.reset-size1.size6{font-size:2em}.katex .fontsize-ensurer.reset-size1.size7,.katex .sizing.reset-size1.size7{font-size:2.4em}.katex .fontsize-ensurer.reset-size1.size8,.katex .sizing.reset-size1.size8{font-size:2.88em}.katex .fontsize-ensurer.reset-size1.size9,.katex .sizing.reset-size1.size9{font-size:3.456em}.katex .fontsize-ensurer.reset-size1.size10,.katex .sizing.reset-size1.size10{font-size:4.148em}.katex .fontsize-ensurer.reset-size1.size11,.katex .sizing.reset-size1.size11{font-size:4.976em}.katex .fontsize-ensurer.reset-size2.size1,.katex .sizing.reset-size2.size1{font-size:.83333333em}.katex .fontsize-ensurer.reset-size2.size2,.katex .sizing.reset-size2.size2{font-size:1em}.katex .fontsize-ensurer.reset-size2.size3,.katex .sizing.reset-size2.size3{font-size:1.16666667em}.katex .fontsize-ensurer.reset-size2.size4,.katex .sizing.reset-size2.size4{font-size:1.33333333em}.katex .fontsize-ensurer.reset-size2.size5,.katex .sizing.reset-size2.size5{font-size:1.5em}.katex .fontsize-ensurer.reset-size2.size6,.katex .sizing.reset-size2.size6{font-size:1.66666667em}.katex .fontsize-ensurer.reset-size2.size7,.katex .sizing.reset-size2.size7{font-size:2em}.katex .fontsize-ensurer.reset-size2.size8,.katex .sizing.reset-size2.size8{font-size:2.4em}.katex .fontsize-ensurer.reset-size2.size9,.katex .sizing.reset-size2.size9{font-size:2.88em}.katex .fontsize-ensurer.reset-size2.size10,.katex .sizing.reset-size2.size10{font-size:3.45666667em}.katex .fontsize-ensurer.reset-size2.size11,.katex .sizing.reset-size2.size11{font-size:4.14666667em}.katex .fontsize-ensurer.reset-size3.size1,.katex .sizing.reset-size3.size1{font-size:.71428571em}.katex .fontsize-ensurer.reset-size3.size2,.katex .sizing.reset-size3.size2{font-size:.85714286em}.katex .fontsize-ensurer.reset-size3.size3,.katex .sizing.reset-size3.size3{font-size:1em}.katex .fontsize-ensurer.reset-size3.size4,.katex .sizing.reset-size3.size4{font-size:1.14285714em}.katex .fontsize-ensurer.reset-size3.size5,.katex .sizing.reset-size3.size5{font-size:1.28571429em}.katex .fontsize-ensurer.reset-size3.size6,.katex .sizing.reset-size3.size6{font-size:1.42857143em}.katex .fontsize-ensurer.reset-size3.size7,.katex .sizing.reset-size3.size7{font-size:1.71428571em}.katex .fontsize-ensurer.reset-size3.size8,.katex .sizing.reset-size3.size8{font-size:2.05714286em}.katex .fontsize-ensurer.reset-size3.size9,.katex .sizing.reset-size3.size9{font-size:2.46857143em}.katex .fontsize-ensurer.reset-size3.size10,.katex .sizing.reset-size3.size10{font-size:2.96285714em}.katex .fontsize-ensurer.reset-size3.size11,.katex .sizing.reset-size3.size11{font-size:3.55428571em}.katex .fontsize-ensurer.reset-size4.size1,.katex .sizing.reset-size4.size1{font-size:.625em}.katex .fontsize-ensurer.reset-size4.size2,.katex .sizing.reset-size4.size2{font-size:.75em}.katex .fontsize-ensurer.reset-size4.size3,.katex .sizing.reset-size4.size3{font-size:.875em}.katex .fontsize-ensurer.reset-size4.size4,.katex .sizing.reset-size4.size4{font-size:1em}.katex .fontsize-ensurer.reset-size4.size5,.katex .sizing.reset-size4.size5{font-size:1.125em}.katex .fontsize-ensurer.reset-size4.size6,.katex .sizing.reset-size4.size6{font-size:1.25em}.katex .fontsize-ensurer.reset-size4.size7,.katex .sizing.reset-size4.size7{font-size:1.5em}.katex .fontsize-ensurer.reset-size4.size8,.katex .sizing.reset-size4.size8{font-size:1.8em}.katex .fontsize-ensurer.reset-size4.size9,.katex .sizing.reset-size4.size9{font-size:2.16em}.katex .fontsize-ensurer.reset-size4.size10,.katex .sizing.reset-size4.size10{font-size:2.5925em}.katex .fontsize-ensurer.reset-size4.size11,.katex .sizing.reset-size4.size11{font-size:3.11em}.katex .fontsize-ensurer.reset-size5.size1,.katex .sizing.reset-size5.size1{font-size:.55555556em}.katex .fontsize-ensurer.reset-size5.size2,.katex .sizing.reset-size5.size2{font-size:.66666667em}.katex .fontsize-ensurer.reset-size5.size3,.katex .sizing.reset-size5.size3{font-size:.77777778em}.katex .fontsize-ensurer.reset-size5.size4,.katex .sizing.reset-size5.size4{font-size:.88888889em}.katex .fontsize-ensurer.reset-size5.size5,.katex .sizing.reset-size5.size5{font-size:1em}.katex .fontsize-ensurer.reset-size5.size6,.katex .sizing.reset-size5.size6{font-size:1.11111111em}.katex .fontsize-ensurer.reset-size5.size7,.katex .sizing.reset-size5.size7{font-size:1.33333333em}.katex .fontsize-ensurer.reset-size5.size8,.katex .sizing.reset-size5.size8{font-size:1.6em}.katex .fontsize-ensurer.reset-size5.size9,.katex .sizing.reset-size5.size9{font-size:1.92em}.katex .fontsize-ensurer.reset-size5.size10,.katex .sizing.reset-size5.size10{font-size:2.30444444em}.katex .fontsize-ensurer.reset-size5.size11,.katex .sizing.reset-size5.size11{font-size:2.76444444em}.katex .fontsize-ensurer.reset-size6.size1,.katex .sizing.reset-size6.size1{font-size:.5em}.katex .fontsize-ensurer.reset-size6.size2,.katex .sizing.reset-size6.size2{font-size:.6em}.katex .fontsize-ensurer.reset-size6.size3,.katex .sizing.reset-size6.size3{font-size:.7em}.katex .fontsize-ensurer.reset-size6.size4,.katex .sizing.reset-size6.size4{font-size:.8em}.katex .fontsize-ensurer.reset-size6.size5,.katex .sizing.reset-size6.size5{font-size:.9em}.katex .fontsize-ensurer.reset-size6.size6,.katex .sizing.reset-size6.size6{font-size:1em}.katex .fontsize-ensurer.reset-size6.size7,.katex .sizing.reset-size6.size7{font-size:1.2em}.katex .fontsize-ensurer.reset-size6.size8,.katex .sizing.reset-size6.size8{font-size:1.44em}.katex .fontsize-ensurer.reset-size6.size9,.katex .sizing.reset-size6.size9{font-size:1.728em}.katex .fontsize-ensurer.reset-size6.size10,.katex .sizing.reset-size6.size10{font-size:2.074em}.katex .fontsize-ensurer.reset-size6.size11,.katex .sizing.reset-size6.size11{font-size:2.488em}.katex .fontsize-ensurer.reset-size7.size1,.katex .sizing.reset-size7.size1{font-size:.41666667em}.katex .fontsize-ensurer.reset-size7.size2,.katex .sizing.reset-size7.size2{font-size:.5em}.katex .fontsize-ensurer.reset-size7.size3,.katex .sizing.reset-size7.size3{font-size:.58333333em}.katex .fontsize-ensurer.reset-size7.size4,.katex .sizing.reset-size7.size4{font-size:.66666667em}.katex .fontsize-ensurer.reset-size7.size5,.katex .sizing.reset-size7.size5{font-size:.75em}.katex .fontsize-ensurer.reset-size7.size6,.katex .sizing.reset-size7.size6{font-size:.83333333em}.katex .fontsize-ensurer.reset-size7.size7,.katex .sizing.reset-size7.size7{font-size:1em}.katex .fontsize-ensurer.reset-size7.size8,.katex .sizing.reset-size7.size8{font-size:1.2em}.katex .fontsize-ensurer.reset-size7.size9,.katex .sizing.reset-size7.size9{font-size:1.44em}.katex .fontsize-ensurer.reset-size7.size10,.katex .sizing.reset-size7.size10{font-size:1.72833333em}.katex .fontsize-ensurer.reset-size7.size11,.katex .sizing.reset-size7.size11{font-size:2.07333333em}.katex .fontsize-ensurer.reset-size8.size1,.katex .sizing.reset-size8.size1{font-size:.34722222em}.katex .fontsize-ensurer.reset-size8.size2,.katex .sizing.reset-size8.size2{font-size:.41666667em}.katex .fontsize-ensurer.reset-size8.size3,.katex .sizing.reset-size8.size3{font-size:.48611111em}.katex .fontsize-ensurer.reset-size8.size4,.katex .sizing.reset-size8.size4{font-size:.55555556em}.katex .fontsize-ensurer.reset-size8.size5,.katex .sizing.reset-size8.size5{font-size:.625em}.katex .fontsize-ensurer.reset-size8.size6,.katex .sizing.reset-size8.size6{font-size:.69444444em}.katex .fontsize-ensurer.reset-size8.size7,.katex .sizing.reset-size8.size7{font-size:.83333333em}.katex .fontsize-ensurer.reset-size8.size8,.katex .sizing.reset-size8.size8{font-size:1em}.katex .fontsize-ensurer.reset-size8.size9,.katex .sizing.reset-size8.size9{font-size:1.2em}.katex .fontsize-ensurer.reset-size8.size10,.katex .sizing.reset-size8.size10{font-size:1.44027778em}.katex .fontsize-ensurer.reset-size8.size11,.katex .sizing.reset-size8.size11{font-size:1.72777778em}.katex .fontsize-ensurer.reset-size9.size1,.katex .sizing.reset-size9.size1{font-size:.28935185em}.katex .fontsize-ensurer.reset-size9.size2,.katex .sizing.reset-size9.size2{font-size:.34722222em}.katex .fontsize-ensurer.reset-size9.size3,.katex .sizing.reset-size9.size3{font-size:.40509259em}.katex .fontsize-ensurer.reset-size9.size4,.katex .sizing.reset-size9.size4{font-size:.46296296em}.katex .fontsize-ensurer.reset-size9.size5,.katex .sizing.reset-size9.size5{font-size:.52083333em}.katex .fontsize-ensurer.reset-size9.size6,.katex .sizing.reset-size9.size6{font-size:.5787037em}.katex .fontsize-ensurer.reset-size9.size7,.katex .sizing.reset-size9.size7{font-size:.69444444em}.katex .fontsize-ensurer.reset-size9.size8,.katex .sizing.reset-size9.size8{font-size:.83333333em}.katex .fontsize-ensurer.reset-size9.size9,.katex .sizing.reset-size9.size9{font-size:1em}.katex .fontsize-ensurer.reset-size9.size10,.katex .sizing.reset-size9.size10{font-size:1.20023148em}.katex .fontsize-ensurer.reset-size9.size11,.katex .sizing.reset-size9.size11{font-size:1.43981481em}.katex .fontsize-ensurer.reset-size10.size1,.katex .sizing.reset-size10.size1{font-size:.24108004em}.katex .fontsize-ensurer.reset-size10.size2,.katex .sizing.reset-size10.size2{font-size:.28929605em}.katex .fontsize-ensurer.reset-size10.size3,.katex .sizing.reset-size10.size3{font-size:.33751205em}.katex .fontsize-ensurer.reset-size10.size4,.katex .sizing.reset-size10.size4{font-size:.38572806em}.katex .fontsize-ensurer.reset-size10.size5,.katex .sizing.reset-size10.size5{font-size:.43394407em}.katex .fontsize-ensurer.reset-size10.size6,.katex .sizing.reset-size10.size6{font-size:.48216008em}.katex .fontsize-ensurer.reset-size10.size7,.katex .sizing.reset-size10.size7{font-size:.57859209em}.katex .fontsize-ensurer.reset-size10.size8,.katex .sizing.reset-size10.size8{font-size:.69431051em}.katex .fontsize-ensurer.reset-size10.size9,.katex .sizing.reset-size10.size9{font-size:.83317261em}.katex .fontsize-ensurer.reset-size10.size10,.katex .sizing.reset-size10.size10{font-size:1em}.katex .fontsize-ensurer.reset-size10.size11,.katex .sizing.reset-size10.size11{font-size:1.19961427em}.katex .fontsize-ensurer.reset-size11.size1,.katex .sizing.reset-size11.size1{font-size:.20096463em}.katex .fontsize-ensurer.reset-size11.size2,.katex .sizing.reset-size11.size2{font-size:.24115756em}.katex .fontsize-ensurer.reset-size11.size3,.katex .sizing.reset-size11.size3{font-size:.28135048em}.katex .fontsize-ensurer.reset-size11.size4,.katex .sizing.reset-size11.size4{font-size:.32154341em}.katex .fontsize-ensurer.reset-size11.size5,.katex .sizing.reset-size11.size5{font-size:.36173633em}.katex .fontsize-ensurer.reset-size11.size6,.katex .sizing.reset-size11.size6{font-size:.40192926em}.katex .fontsize-ensurer.reset-size11.size7,.katex .sizing.reset-size11.size7{font-size:.48231511em}.katex .fontsize-ensurer.reset-size11.size8,.katex .sizing.reset-size11.size8{font-size:.57877814em}.katex .fontsize-ensurer.reset-size11.size9,.katex .sizing.reset-size11.size9{font-size:.69453376em}.katex .fontsize-ensurer.reset-size11.size10,.katex .sizing.reset-size11.size10{font-size:.83360129em}.katex .fontsize-ensurer.reset-size11.size11,.katex .sizing.reset-size11.size11{font-size:1em}.katex .delimsizing.size1{font-family:KaTeX_Size1}.katex .delimsizing.size2{font-family:KaTeX_Size2}.katex .delimsizing.size3{font-family:KaTeX_Size3}.katex .delimsizing.size4{font-family:KaTeX_Size4}.katex .delimsizing.mult .delim-size1>span{font-family:KaTeX_Size1}.katex .delimsizing.mult .delim-size4>span{font-family:KaTeX_Size4}.katex .nulldelimiter{display:inline-block;width:.12em}.katex .delimcenter,.katex .op-symbol{position:relative}.katex .op-symbol.small-op{font-family:KaTeX_Size1}.katex .op-symbol.large-op{font-family:KaTeX_Size2}.katex .accent>.vlist-t,.katex .op-limits>.vlist-t{text-align:center}.katex .accent .accent-body{position:relative}.katex .accent .accent-body:not(.accent-full){width:0}.katex .overlay{display:block}.katex .mtable .vertical-separator{display:inline-block;min-width:1px}.katex .mtable .arraycolsep{display:inline-block}.katex .mtable .col-align-c>.vlist-t{text-align:center}.katex .mtable .col-align-l>.vlist-t{text-align:left}.katex .mtable .col-align-r>.vlist-t{text-align:right}.katex .svg-align{text-align:left}.katex svg{fill:currentColor;stroke:currentColor;fill-rule:nonzero;fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke-miterlimit:4;stroke-dasharray:none;stroke-dashoffset:0;stroke-opacity:1;display:block;height:inherit;position:absolute;width:100%}.katex svg path{stroke:none}.katex img{border-style:none;max-height:none;max-width:none;min-height:0;min-width:0}.katex .stretchy{display:block;overflow:hidden;position:relative;width:100%}.katex .stretchy:after,.katex .stretchy:before{content:""}.katex .hide-tail{overflow:hidden;position:relative;width:100%}.katex .halfarrow-left{left:0;overflow:hidden;position:absolute;width:50.2%}.katex .halfarrow-right{overflow:hidden;position:absolute;right:0;width:50.2%}.katex .brace-left{left:0;overflow:hidden;position:absolute;width:25.1%}.katex .brace-center{left:25%;overflow:hidden;position:absolute;width:50%}.katex .brace-right{overflow:hidden;position:absolute;right:0;width:25.1%}.katex .x-arrow-pad{padding:0 .5em}.katex .cd-arrow-pad{padding:0 .55556em 0 .27778em}.katex .mover,.katex .munder,.katex .x-arrow{text-align:center}.katex .boxpad{padding:0 .3em}.katex .fbox,.katex .fcolorbox{border:.04em solid;box-sizing:border-box}.katex .cancel-pad{padding:0 .2em}.katex .cancel-lap{margin-left:-.2em;margin-right:-.2em}.katex .sout{border-bottom-style:solid;border-bottom-width:.08em}.katex .angl{border-right:.049em solid;border-top:.049em solid;box-sizing:border-box;margin-right:.03889em}.katex .anglpad{padding:0 .03889em}.katex .eqn-num:before{content:"(" counter(katexEqnNo) ")";counter-increment:katexEqnNo}.katex .mml-eqn-num:before{content:"(" counter(mmlEqnNo) ")";counter-increment:mmlEqnNo}.katex .mtr-glue{width:50%}.katex .cd-vert-arrow{display:inline-block;position:relative}.katex .cd-label-left{display:inline-block;position:absolute;right:calc(50% + .3em);text-align:left}.katex .cd-label-right{display:inline-block;left:calc(50% + .3em);position:absolute;text-align:right}.katex-display{display:block;margin:1em 0;text-align:center}.katex-display>.katex{display:block;text-align:center;white-space:nowrap}.katex-display>.katex>.katex-html{display:block;position:relative}.katex-display>.katex>.katex-html>.tag{position:absolute;right:0}.katex-display.leqno>.katex>.katex-html>.tag{left:0;right:auto}.katex-display.fleqn>.katex{padding-left:2em;text-align:left}body{counter-reset:katexEqnNo mmlEqnNo}

/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.katex-error {
	color: var(--vscode-editorError-foreground);
}

/* From extension ms-toolsai.jupyter */
/* These classnames are inherited from bootstrap, but are present in most notebook renderers */

.alert {
    width: auto;
    padding: 1em;
    margin-top: 1em;
    margin-bottom: 1em;
}
.alert > *:last-child {
    margin-bottom: 0;
}
#preview > .alert:last-child {
    /* Prevent this being set to zero by the default notebook stylesheet */
    padding-bottom: 1em;
}

.alert-success {
    /* Note there is no suitable color available, so we just copy "info" */
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-info {
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-warning {
    background-color: var(--theme-warning-background);
    color: var(--theme-warning-foreground);
}
.alert-danger {
    background-color: var(--theme-error-background);
    color: var(--theme-error-foreground);
}

</style>
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.10.2/dist/katex.min.css" integrity="sha384-yFRtMMDnQtDRO8rLpMIKrtPCD5jdktao2TV19YiZYWMDkUR5GQZR/NOVTdquEx1j" crossorigin="anonymous">
<link href="https://cdn.jsdelivr.net/npm/katex-copytex@latest/dist/katex-copytex.min.css" rel="stylesheet" type="text/css">
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/markdown.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/highlight.css">
<style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif;
                font-size: 14px;
                line-height: 1.6;
            }
        </style>
        <style>
.task-list-item { list-style-type: none; } .task-list-item-checkbox { margin-left: -20px; vertical-align: middle; }
</style>
        
    </head>
    <body class="vscode-body vscode-light">
        <h1 id="metabarcoding-workshop">Metabarcoding workshop</h1>
<p><em>29th July 2022
Weymouth</em></p>
<p><strong>Author:</strong> Nicola Coyle
Based in part on the Dada2 tutorial: <a href="https://benjjneb.github.io/dada2/tutorial.html">https://benjjneb.github.io/dada2/tutorial.html</a>
Further resources can be found here: <a href="https://metabarcoding-at-cefas-tutorial.readthedocs.io/en/latest/index.html">https://metabarcoding-at-cefas-tutorial.readthedocs.io/en/latest/index.html</a></p>
<p>If you would like to try this on your own data you can use the script supplied here: <a href="https://github.com/nmc97/metabarcoding_at_cefas_tutorial/blob/main/scripts/run_dada2.R">https://github.com/nmc97/metabarcoding_at_cefas_tutorial/blob/main/scripts/run_dada2.R</a></p>
<h2 id="1-set-up">1. Set up</h2>
<p>First lets set up a directory to do our work in. We will call is metabarcoding_ws and place it in our home directory (<code>~/</code> or <code>/home/$USER</code>).</p>
<p>Next we will download reference databases for classifying our barcodes. in this example we will be using the SILVA database.</p>
<p>Then we will download our data and get started!</p>
<p><strong>Some of the basic commands we will use:</strong>
<code>mkdir /path/to/directory</code>: This makes a directory
<code>cd /path/to/directory</code>: This changes directory
<code>mv /path/to/file /path/to/new_file_location</code>: This will move a file or folder from one place to another. You can use this to change the file name.</p>
<h3 id="11-make-a-directory-to-run-a-the-analysis">1.1 Make a directory to run a the analysis</h3>
<pre><code class="language-bash"><div>mkdir ~/metabarcoding_ws <span class="hljs-comment"># makes a directory metabarcoding_ws in the home directory `~/`</span>
mkdir ~/metabarcoding_ws/data <span class="hljs-comment"># make a directory to save data</span>
mkdir ~/metabarcoding_ws/data/16S <span class="hljs-comment"># specify a data directory for 16S data</span>
mkdir ~/metabarcoding_ws/outputs <span class="hljs-comment"># set directory to save outputs</span>
mkdir ~/metabarcoding_ws/db <span class="hljs-comment"># optionally put databases into a directory within workshop directory</span>
</div></code></pre>
<h3 id="12-download-reference-databases">1.2 <a href="https://metabarcoding-at-cefas-tutorial.readthedocs.io/en/latest/dada2.html#download-silva-datasets-curated-for-dada2">Download reference databases</a></h3>
<p>Download Silva dataset for the tutorial (~ 1 mins 30 seconds, 131 Mb):</p>
<pre><code><code><div># move to the directory where we will store the databases using `cd` (change directory)
cd metabarcoding_ws/db

# wget grabs the file from the internet and downloads into the current directory
wget https://zenodo.org/record/4587955/files/silva_nr99_v138.1_train_set.fa.gz?download=1

# rename the file to remove &quot;?download=1&quot;. Uses command `mv`
mv silva_nr99_v138.1_train_set.fa.gz?download=1 silva_nr99_v138.1_train_set.fa.gz
</div></code></code></pre>
<p>Download Silva <em>species</em> dataset for the tutorial (~ 30 seconds, 76 M):</p>
<pre><code><code><div>cd metabarcoding_ws/db # set a directory to store the data
wget https://zenodo.org/record/4587955/files/silva_species_assignment_v138.1.fa.gz?download=1 # grabs the file from the internet and downloads into the current directory
mv silva_species_assignment_v138.1.fa.gz?download=1 silva_species_assignment_v138.1.fa.gz # renames the file to remove &quot;?download=1&quot;
</div></code></code></pre>
<h3 id="13-download-test-data">1.3 Download test data</h3>
<p>Here we will use a dataset produced in collaboration with Cefas:</p>
<p>Jamie McMurtrie, Shayma Alathari, Dominique L. Chaput, David Bass, Camerson Ghambi, Joseph Nagoli, Jérôme Delamare-Deboutteville, Chadag Vishnumurthy Mohan, Joanne Cable, Ben Temperton, Charles R. Tyler, <em>Relationships between pond water and tilapia skin microbiomes in aquaculture ponds in Malawi</em>, Aquaculture, Volume 558, 2022, 738367, ISSN 0044-8486, <a href="https://doi.org/10.1016/j.aquaculture.2022.738367">https://doi.org/10.1016/j.aquaculture.2022.738367</a>.</p>
<p>NCBI project code: PRJEB46984</p>
<p>Code for the project is available here: <a href="https://github.com/jamiemcm/Malawi_Tilapia_Microbiomes">https://github.com/jamiemcm/Malawi_Tilapia_Microbiomes</a>.</p>
<p>There are two ways to get the data added to your POD:</p>
<p><strong>Option 1: Take it from my POD (recommended)</strong></p>
<pre><code class="language-bash"><div><span class="hljs-comment"># this will take my fastq folder containing read files and place it in your 16S directory</span>
rsync -ravz /home/nc07/metabarcoding_ws/data/16S/fastq/ ~/metabarcoding_ws/data/16S/
</div></code></pre>
<p><strong>Option 2: Download from SRA (long)</strong>
Uses SRA-tools: <a href="https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump">https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump</a></p>
<pre><code class="language-bash"><div><span class="hljs-comment"># make and activate a conda env for sra-tools</span>
mamba create -n sra-tools
conda activate sra-tools

mkdir ~/metabarcoding_ws/data/16S
mkdir ~/metabarcoding_ws/data/16S/fastq

<span class="hljs-built_in">cd</span> ~/metabarcoding_ws/data/16S

<span class="hljs-comment"># download list of sample names and accession</span>
wget https://raw.githubusercontent.com/nmc97/metabarcoding_at_cefas_tutorial/main/scripts/workshop_1/data/names_16S.txt

<span class="hljs-comment"># prefetch data - uses file names_16S.txt to find the data (accession: q) download them (prefetch) and then rename the files (name: p)</span>
<span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> p q; <span class="hljs-keyword">do</span>
<span class="hljs-built_in">echo</span> <span class="hljs-variable">$p</span>
time prefetch <span class="hljs-variable">$q</span>
<span class="hljs-keyword">done</span> &lt; names_16S.txt

<span class="hljs-comment"># extract fastq files</span>
<span class="hljs-keyword">while</span> <span class="hljs-built_in">read</span> p q; <span class="hljs-keyword">do</span>
<span class="hljs-built_in">echo</span> <span class="hljs-variable">$p</span>
ls <span class="hljs-variable">$q</span>/<span class="hljs-variable">$q</span>.sra
time fasterq-dump --split-files <span class="hljs-variable">$q</span> -o <span class="hljs-variable">$q</span>/<span class="hljs-variable">$p</span>
<span class="hljs-keyword">done</span> &lt; names_16S.txt

<span class="hljs-comment"># move files and zip</span>
mv ~/metabarcoding_ws/16S/*/*.fastq ~/metabarcoding_ws/16S/fastq
gzip ~/metabarcoding_ws/16S/fastq/*.fastq
</div></code></pre>
<h3 id="14-set-up-conda-environments">1.4 Set up conda environments</h3>
<p>Set up Fastqc env:</p>
<pre><code class="language-bash"><div><span class="hljs-comment"># create an environment called fastqc and install fastqc and multiqc</span>
mamba create -n fastqc fastqc MultiQC
</div></code></pre>
<p>If this is your first time creating a conda environment, or it has been a while, here's a breakdown of what just happened:</p>
<p>&quot;mamba&quot;: calls the program mamba
&quot;create&quot;: tells mamba you want to create a new environment
&quot;-n fastqc&quot;: names the new environment fastqc
&quot;fastqc&quot;: installs the program named fastqc
&quot;MultiQC&quot;: installs the program MultiQC</p>
<p>You can list as many programs as you want after naming the environment and Mamba will attempt to install them.</p>
<p>However, it is good practice to keep individual environments for specific programs. This is because when you add a new program to environment, it may require different dependencies than programs already in the environment. Thus, the dependencies could clash and cause one or many programs so stop working. Here we will keep the number of programs in each environment low, and try to group them by task. Fastqc and MultiQC are frequently used together and work well in and envrionment.</p>
<p>In the next environment we need to set up R using the package r-essentials. Thn we can install dada2, which runs in R.</p>
<p>Set up dada2 env:</p>
<pre><code class="language-bash"><div>mamba create -n dada2 r-essentials <span class="hljs-comment"># setup a new environment and install r-essentials</span>
conda activate dada2 <span class="hljs-comment"># activate the new environment</span>
mamba  install bioconductor-dada2 <span class="hljs-comment"># install dada2</span>
</div></code></pre>
<p>Lastly we will install dadasit2, which is a wrapper program for dada2. Meaning, it bundled up code for you to run dada2 using a single line of code in the command-line. This could be useful if you have data that you want to quickly analyse which isn't too complex, requiring more manual coding work.</p>
<p>Set up dadaist2 env:</p>
<pre><code class="language-bash"><div>mamba create -n dadaist2
conda activate dadaist2
mamba install -c conda-forge -c bioconda dadaist2
mamba install bioconductor-dada2=1.20
mamba install -c conda-forge pyyaml <span class="hljs-comment"># optional: needed to run dadaist2-mqc-report</span>
</div></code></pre>
<h2 id="2-getting-our-pod-session-set-up">2. Getting our POD session set up</h2>
<h3 id="21-start-a-screen-session">2.1 Start a screen session</h3>
<p>When you are logged in to POD, and your connection drops, you will loose all the progress you had made in that terminal.</p>
<p>Screen allows you to set up a session or a &quot;screen&quot; in which to do your work. You can shut the screen down and it will continue to run in the background. We will set one up for this session:</p>
<pre><code class="language-bash"><div>screen -S meta
</div></code></pre>
<p>You can leave the screen by pressing <code>ctrl + a + d</code></p>
<p>Restart the screen by typing <code>screen -r meta</code></p>
<h3 id="22-start-an-interactive-session-for-the-day">2.2 Start an interactive session for the day:</h3>
<p>Where we are all working now, there isn't much computing power available. In POD we will need to each ask for more by starting an interactive session.</p>
<p>Here, you can change the S30 to B30 if there is no more space left on S30.</p>
<p>See <a href="https://cefas.sharepoint.com/sites/PODHighPerformanceComputingUsers/SitePages/Running%20Jobs%20on%20MT2.aspx">here</a> for more information.</p>
<pre><code class="language-bash"><div>msub -I -q S30 -l procs=8,walltime=7:00:00 <span class="hljs-comment"># change depending on what you think you need.</span>
</div></code></pre>
<h2 id="3-quality-control">3. Quality control</h2>
<p>First thing we need to do is check the quality of the data. We will run fastqc which will assess all the fastq files filled with reads. It will tell us how many reads we have and what they look like in regards to length and quality.</p>
<h2 id="31-run-fastqc-and-multiqc">3.1 Run <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">Fastqc</a> and <a href="https://www.bing.com/search?q=multiqc&amp;cvid=41c889025cf2474eae8615972425b160&amp;aqs=edge..69i57j0l8j69i11004.1847j0j4&amp;FORM=ANAB01&amp;PC=U531">Multiqc</a></h2>
<p>The first thing we will do is activate the conda envrionment we made earlier called &quot;fastqc&quot;. This environment contains programs fastqc and MultiQC.</p>
<p>Then we will set some paths. This means that we will set two variables <code>$inder</code> and <code>$outdir</code> which contain the paths on the files system. Variables in bash are denoted using a <code>$</code>. The first path <code>$inder</code> will contain the path to the input data for the experiment. We put this data in he directory (folder) called <code>~/metabarcoding_ws/data/16S/fastq</code>, so this will be set as our <code>$inder</code> variable.</p>
<p>If the output directory for fastqc doesn't exist it will not be happy, so we need to make it. To make a directory we use the command <code>mkdir</code>.</p>
<p>After we make this directory we run fastqc.</p>
<p>Lastly we will run MultiQC, which collates all the outputs from fastqc and puts them into one report.</p>
<p>Okay, let's try it:</p>
<pre><code class="language-bash"><div><span class="hljs-comment">#Activate conda environment:</span>
	conda activate fastqc

<span class="hljs-comment">#Set paths</span>
	indir=~/metabarcoding_ws/data/16S/fastq <span class="hljs-comment"># this is where our input data lives</span>
	outdir=~/metabarcoding_ws/outputs <span class="hljs-comment"># this is where all our outputs will go</span>

<span class="hljs-comment">#Make a directory to store fastqc files</span>
	mkdir <span class="hljs-variable">$outdir</span>/fastqc_out

</div></code></pre>
<p><em>Aside:</em>
Lets check that that worked by listing all the files in our output directory.</p>
<p><code>ls $outdir</code>: ls is a command to list files. Here it will list all the files in <code>$outdir</code>.</p>
<p>If it gives <code>fastqc_ws</code> as an output on the terminal, then we successfully created that directory.</p>
<p>Now lets run FastQC:</p>
<pre><code class="language-bash"><div><span class="hljs-comment">#Run fastqc:</span>
	fastqc <span class="hljs-variable">$indir</span>/*.fastq.gz -t 8 -o <span class="hljs-variable">$outdir</span>/fastqc_out
<span class="hljs-comment"># runs fastqc on all fastq.gz n current directory. -t threads to use</span>

<span class="hljs-comment">#Run multiqc on fastqc output folder.</span>
<span class="hljs-comment">#(It automatically detects fastqc outputs)</span>
<span class="hljs-comment">#change `--title` if you wish</span>

	multiqc <span class="hljs-variable">$outdir</span>/fastqc_out/* -o <span class="hljs-variable">$outdir</span>/fastqc_out/multiqc_fastqc --title fastqc
</div></code></pre>
<p>This gives a message at the end:</p>
<pre><code class="language-bash"><div>|           multiqc | 8 flat-image plots used <span class="hljs-keyword">in</span> the report due to large sample numbers
|           multiqc | To force interactive plots, use the <span class="hljs-string">&#x27;--interactive&#x27;</span> flag. See the documentation.
</div></code></pre>
<p>So lets add the <code>--interactive</code> flag to the command, and change the output name:</p>
<pre><code class="language-bash"><div><span class="hljs-comment"># force interative if there are lots of files (it will tell you if it wrote flat files instead)</span>
	multiqc <span class="hljs-variable">$outdir</span>/fastqc_out/* -o <span class="hljs-variable">$outdir</span>/fastqc_out/multiqc_fastqc --interactive --title fastqc_interactive
</div></code></pre>
<p>Now we will check the quality of the data and decide how we would like to trim and filter them in the next step.</p>
<p>You can do this in MobaXterm by navigating to the MultiQC output directory and selecting the <code>.html</code> in the file.</p>
<p>If we would like to, we can find out the full path to this directory by using <code>echo</code>:</p>
<p><code>echo $outdir/fastqc_out/multiqc_fastqc</code>: this will print the output of <code>$outdir/fastqc_out/multiqc_fastqc</code> to the command line with the variable <code>$outdir</code> filled in.</p>
<p>Once you find the <code>html</code> file, right click and select <code>open with</code>. Then select a browser.</p>
<ul class="contains-task-list">
<li class="task-list-item">
<p><input class="task-list-item-checkbox" disabled="" type="checkbox"> Question: What do you notice about the reads?</p>
</li>
<li class="task-list-item">
<p><input class="task-list-item-checkbox" disabled="" type="checkbox"> ## 4. Run <a href="https://github.com/nmc97/metabarcoding_at_cefas_tutorial/blob/main/scripts/run_dada2.R">dada2 script</a> or <a href="https://github.com/nmc97/metabarcoding_at_cefas_tutorial/blob/main/scripts/run_dada2.R">looped dada2 script</a></p>
</li>
<li class="task-list-item">
<p><input class="task-list-item-checkbox" disabled="" type="checkbox"> ## 4.1 Activate dada2 environment and start an R session</p>
</li>
</ul>
<p>Firstly, lets deactivate the fastqc environment now that we are done with it:</p>
<pre><code class="language-bash"><div><span class="hljs-comment"># deactivate fastqc conda environment</span>
conda deactivate fastqc
</div></code></pre>
<p>Next we will activate out dada2 environment, and start an R session. We do this in the terminal by typing R.</p>
<pre><code class="language-bash"><div><span class="hljs-comment"># activate dada2 conda environment</span>
conda activate dada2

<span class="hljs-comment"># start R session</span>
R
</div></code></pre>
<ul class="contains-task-list">
<li class="task-list-item"><input class="task-list-item-checkbox" disabled="" type="checkbox"> ## 4.2 Set up R:</li>
</ul>
<p>Frist thing to do in R is load the dada2 library.</p>
<p>Next we will do the same as above and set some variables (outpath, path). Note that in R we don't need to indicate something is a variable using <code>$</code> as in bash.</p>
<pre><code class="language-R"><div><span class="hljs-comment">#=========================#</span>
<span class="hljs-comment"># load libraries</span>
<span class="hljs-comment"># we only need dada2</span>
<span class="hljs-comment">#=========================#</span>

library(dada2)

<span class="hljs-comment">#=========================#</span>
<span class="hljs-comment"># setup file paths</span>
<span class="hljs-comment">#=========================#</span>

<span class="hljs-comment"># set path to output directory outpath</span>
outpath &lt;- <span class="hljs-string">&quot;~/metabarcoding_ws/outputs/dada2/&quot;</span>
<span class="hljs-comment"># set path to data directory</span>
path &lt;- <span class="hljs-string">&quot;~/metabarcoding_ws/data/16S/fastq&quot;</span>

<span class="hljs-comment"># check if the output folder exists.</span>
<span class="hljs-comment"># Use this if statement to make an output folder if doesn&#x27;t exist:</span>
<span class="hljs-keyword">if</span> (file.exists(outpath)) {
 cat(<span class="hljs-string">&quot;The folder already exists&quot;</span>)
} <span class="hljs-keyword">else</span> {
 dir.create(outpath)
 cat(paste(<span class="hljs-string">&quot;making a directory:&quot;</span>, outpath, <span class="hljs-string">&quot;\n&quot;</span>))
}

<span class="hljs-comment"># check input files exist in read file directory</span>
<span class="hljs-comment"># list.files will list all the files in path</span>
list.files(path)
</div></code></pre>
<p>dada2 uses lists of file names, so we need to produce them. We will produce a list of forward read files <code>fnFs</code> and reverse reads <code>fnrs</code>.</p>
<blockquote>
<p><em>Note:</em> if your files end in something different to <code>_1.fastq.gz</code> or <code>_2.fastq.gz</code> you will need to change the code to match your data.
In particular: <code>pattern=&quot;_1.fastq&quot;</code> and <code>pattern=&quot;_2.fastq&quot;</code></p>
</blockquote>
<pre><code class="language-R"><div><span class="hljs-comment"># assuming Forward and reverse fastq filenames have format: SAMPLENAME_1.fastq and SAMPLENAME_R2_001.fastq</span>

<span class="hljs-comment"># list forward read files</span>

fnFs &lt;- sort(list.files(path, pattern=<span class="hljs-string">&quot;_1.fastq&quot;</span>, full.names = <span class="hljs-literal">TRUE</span>)) <span class="hljs-comment"># this will find all files otaining &quot;_1.fastq&quot;</span>

<span class="hljs-comment"># list reverse read files</span>

fnRs &lt;- sort(list.files(path, pattern=<span class="hljs-string">&quot;_2.fastq&quot;</span>, full.names = <span class="hljs-literal">TRUE</span>)) <span class="hljs-comment"># this will find all files otaining &quot;_2.fastq&quot;</span>

<span class="hljs-comment"># Extract and keep a list of sample names, assuming filenames have format: SAMPLENAME_XXX.fastq</span>
sample.names &lt;- sapply(strsplit(basename(fnFs), <span class="hljs-string">&quot;_&quot;</span>), `[`, <span class="hljs-number">1</span>)
</div></code></pre>
<ul class="contains-task-list">
<li class="task-list-item"><input class="task-list-item-checkbox" disabled="" type="checkbox"> ## 4.3 Filter and trim:</li>
</ul>
<p>Now that we have set up all the file paths, we can have a second look at the quality of the reads. This time, using dada2.</p>
<p>Then we can decide how we would like to trim and filter them.</p>
<ul class="contains-task-list">
<li class="task-list-item"><input class="task-list-item-checkbox" disabled="" type="checkbox"> ### 4.3.1 Dada2 filter and trim</li>
</ul>
<p>For completeness - we will check read quality in dada2. dada2 has a function <code>plotQualityProfile</code> which can be used to view the read quality of your samples.</p>
<pre><code class="language-R"><div><span class="hljs-comment"># make a pdf of the read quality output plot_read_quality_F.pdf</span>
<span class="hljs-comment"># 12 samples included</span>
pdf(file = paste(outpath,<span class="hljs-string">&quot;/plot_read_quality_F.pdf&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>), <span class="hljs-number">7</span>, <span class="hljs-number">7</span>)
plotQualityProfile(fnFs[<span class="hljs-number">1</span>:<span class="hljs-number">12</span>])
dev.off()

<span class="hljs-comment"># make a pdf of the read quality output plot_read_quality_R.pdf# 12 samples included</span>

pdf(file = paste(outpath,<span class="hljs-string">&quot;/plot_read_quality_R.pdf&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>), <span class="hljs-number">7</span>, <span class="hljs-number">7</span>)
plotQualityProfile(fnRs[<span class="hljs-number">1</span>:<span class="hljs-number">12</span>])
dev.off()
</div></code></pre>
<blockquote>
<p>Note: you will see the following error message:
<code>guides(&lt;scale&gt; = FALSE)</code> is deprecated. Please use <code>guides(&lt;scale&gt; = &quot;none&quot;)</code> instead.
Ignore this.</p>
</blockquote>
<p>Now we will filter and trim. First we will make lists of the outputs files we will produce: <code>filtFs</code> and <code>filtRs</code>.</p>
<pre><code class="language-R"><div><span class="hljs-comment"># list filtered files in subdirectory: called filtered</span>

<span class="hljs-comment"># make a list of filtered output files for forward reads:</span>
filtFs &lt;- file.path(outpath, <span class="hljs-string">&quot;filtered&quot;</span>, paste0(sample.names, <span class="hljs-string">&quot;_F_filt.fastq.gz&quot;</span>)) <span class="hljs-comment"># forward reads</span>
<span class="hljs-built_in">names</span>(filtFs) &lt;- sample.names

<span class="hljs-comment"># make a list of filtered output files for reverse reads:</span>
filtRs &lt;- file.path(outpath, <span class="hljs-string">&quot;filtered&quot;</span>, paste0(sample.names, <span class="hljs-string">&quot;_R_filt.fastq.gz&quot;</span>)) <span class="hljs-comment"># reverse reads</span>
<span class="hljs-built_in">names</span>(filtRs) &lt;- sample.names

</div></code></pre>
<p>The <code>filterAndTrim</code> function</p>
<blockquote>
<p>Parameters to think about are:
<code>trimleft</code>
<code>trunclen</code>
<code>maxN</code>
<code>maxEE</code>
<code>truncQ</code></p>
</blockquote>
<pre><code><code><div># Action required: Change trunclen=c(xx,xx) to match what you want to truncate your forward and reverse reads to. Similarly edit trimLeft=c(xx,xx), maxEE=x and truncQ=x. ##change_me
# https://rdrr.io/bioc/dada2/man/filterAndTrim.html
out &lt;- filterAndTrim(fnFs, filtFs, fnRs, filtRs, trimLeft=c(10,10), truncLen=c(200,160),
              maxN=0, maxEE=c(2,2), truncQ=c(2,2), rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=TRUE
head(out) # check how many reads have been lost in filtering step
# save out file in case environment shuts down
write.table(out,paste(outpath,&quot;/out_save_F1.tsv&quot;,sep=&quot;&quot;),sep=&quot;\t&quot;,row.names=T)
</div></code></code></pre>
<p>At this point we can look at the summary of the trimming step by looking at the varable <code>out</code>. To do this type <code>out</code> into the terminal.</p>
<blockquote>
<p>Or, for ease, print &quot;out&quot; ordered by column two, which is the number of reads left: <code>out[order(out[,2], decreasing=T),]</code></p>
</blockquote>
<ul class="contains-task-list">
<li class="task-list-item"><input class="task-list-item-checkbox" disabled="" type="checkbox"> Questions: Have many reads been lost? Are there any samples with too few reads to proceed with?</li>
</ul>
<p>Next we can make more plots of the read files, this time after they have been trimmed.</p>
<pre><code class="language-R"><div><span class="hljs-comment"># make a pdf of the read quality output plot_read_quality_F.pdf</span>
pdf(file = paste(outpath,<span class="hljs-string">&quot;/plot_read_quality_filtered_F.pdf&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>), <span class="hljs-number">7</span>, <span class="hljs-number">7</span>)
plotQualityProfile(filtFs[<span class="hljs-number">1</span>:<span class="hljs-number">12</span>])
dev.off()

<span class="hljs-comment"># make a pdf of the read quality output plot_read_quality_R.pdf</span>
pdf(file = paste(outpath,<span class="hljs-string">&quot;/plot_read_quality_filtered_R.pdf&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>), <span class="hljs-number">7</span>, <span class="hljs-number">7</span>)
plotQualityProfile(filtRs[<span class="hljs-number">1</span>:<span class="hljs-number">12</span>])
dev.off()
</div></code></pre>
<p>Sometimes files lose all reads at this step. If that happens you need to exclude them from the rest the analysis.</p>
<p>Or, if there are samples with some reads, but you would like to exclude them from the rest of the analysis, you can remove them with the code below.</p>
<p>More specifically, the variable <code>removed_samples</code> is made first, and is populated by a list of samples with zero reads left (<code>which(as.data.frame(out)[,2] ==0</code>). To add samples to this list you need to use <code>c(removed samples, &quot;sample_file_1.fastq.gz&quot;)</code>
e.g.:
<code>removed_samples&lt;- c(removed_samples, &quot;P1a_1.fastq.gz&quot;, &quot;P2a_1.fastq.gz&quot;, &quot;P1e_1.fastq.gz&quot;, &quot;P1c_1.fastq.gz&quot;)</code></p>
<pre><code class="language-R"><div><span class="hljs-comment"># if statement checks if samples need to be removed and following statements remove them</span>
removed_samples&lt;-row.names(as.data.frame(out)[(which(as.data.frame(out)[,<span class="hljs-number">2</span>] ==<span class="hljs-number">0</span>)),])

<span class="hljs-comment"># modify this line to remove samples manually.</span>
<span class="hljs-comment"># here, sample Ts29 has fewer reads than the rest of the dataset so we will remove it</span>
removed_samples&lt;-<span class="hljs-built_in">c</span>(removed_samples,<span class="hljs-string">&quot;Ts29_1.fastq.gz&quot;</span>)

<span class="hljs-comment"># clean up file names</span>
removed_samples.names &lt;- sapply(strsplit(basename(removed_samples), <span class="hljs-string">&quot;_&quot;</span>), `[`, <span class="hljs-number">1</span>) <span class="hljs-comment"># check this works for your data</span>

<span class="hljs-comment"># remove samples from filtFs, filtRs, sample.names</span>
<span class="hljs-keyword">if</span>(<span class="hljs-built_in">length</span>(removed_samples) != <span class="hljs-number">0</span>){
  filtFs&lt;-filtFs[(!(<span class="hljs-built_in">names</span>(filtFs) %in% removed_samples.names))] <span class="hljs-comment"># remove from list of forward filtered read files</span>
  filtRs&lt;-filtRs[(!(<span class="hljs-built_in">names</span>(filtRs) %in% removed_samples.names))] <span class="hljs-comment"># remove from list of reverse filtered read files</span>
  sample.names&lt;-sample.names[!(sample.names) %in% removed_samples.names]
  out&lt;-out[which(!(row.names(out)%in% removed_samples)),]
}

</div></code></pre>
<p>Stop here and check the filtering and trimming step before moving forward!</p>
<ul class="contains-task-list">
<li class="task-list-item"><input class="task-list-item-checkbox" disabled="" type="checkbox"> Do you need to repeat with new parameters?</li>
</ul>
<p>So, if it seems we have lost too many reads, or not enough, or the quality scores are still too low, we need to read try again with new parameters.</p>
<p>We can check the read quality using FastQC again to be sure:</p>
<ul class="contains-task-list">
<li class="task-list-item">
<p><input class="task-list-item-checkbox" disabled="" type="checkbox"> ### 4.3.2 Run <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">Fastqc</a> and <a href="https://www.bing.com/search?q=multiqc&amp;cvid=41c889025cf2474eae8615972425b160&amp;aqs=edge..69i57j0l8j69i11004.1847j0j4&amp;FORM=ANAB01&amp;PC=U531">Multiqc</a> [optional]</p>
</li>
<li class="task-list-item">
<p><input class="task-list-item-checkbox" disabled="" type="checkbox"> #### Open a new command-line session:</p>
</li>
</ul>
<p>So that we don't interrupt the R session we have been working in, let's start a new session.</p>
<p>You can either exit your screen using <code>ctrl + a + d</code>, and start a new one <code>screen -S fastqc</code>.</p>
<p>Or you can start a new session using Mobaxterm.</p>
<pre><code class="language-bash"><div><span class="hljs-comment"># start another interactive session in a different session</span>
	msub -I -q S30 -l procs=8,walltime=0:30:00

<span class="hljs-comment"># Activate conda environment:</span>
	conda activate fastqc

<span class="hljs-comment"># Set paths</span>
	indir=~/metabarcoding_ws/outputs/dada2/filtered
	outdir=~/metabarcoding_ws/outputs/dada2/

<span class="hljs-comment"># Make a directory to store fastqc files</span>
	mkdir <span class="hljs-variable">$outdir</span>/filtered_fastqc_out

<span class="hljs-comment"># Run fastqc:</span>
	fastqc <span class="hljs-variable">$indir</span>/*.fastq.gz -t 8 -o <span class="hljs-variable">$outdir</span>/filtered_fastqc_out
<span class="hljs-comment"># runs fastqc on all fastq.gz n current directory. -t threads to use</span>

<span class="hljs-comment"># Run multiqc on fastqc output folder.</span>
<span class="hljs-comment">#(It automatically detects fastqc outputs)</span>
<span class="hljs-comment"># change `--title` if you wish</span>

	multiqc <span class="hljs-variable">$outdir</span>/filtered_fastqc_out/* -o <span class="hljs-variable">$outdir</span>/filtered_fastqc_out/multiqc_fastqc --title fastqc

<span class="hljs-comment">#Note: force interative if there are lots of files (it will tell you if it wrote flat files instead)</span>
	multiqc <span class="hljs-variable">$outdir</span>/filtered_fastqc_out/* -o <span class="hljs-variable">$outdir</span>/filtered_fastqc_out/multiqc_fastqc --interactive --title fastqc_filtered_interactive
</div></code></pre>
<p>If you are happy to proceed, go back to the R session and start to calculate error rates:</p>
<ul class="contains-task-list">
<li class="task-list-item"><input class="task-list-item-checkbox" disabled="" type="checkbox"> ## 4.4 Error rates</li>
<li class="task-list-item"><input class="task-list-item-checkbox" disabled="" type="checkbox"> #### Back in R:</li>
</ul>
<p>Dada2 relies on error models calculated for every dataset. the function <code>learnErrors()</code> will calculate these and store them in <code>errF</code> and <code>errR</code></p>
<pre><code class="language-R"><div><span class="hljs-comment"># calculate error rates for forward and reverse reads</span>
errF &lt;- learnErrors(filtFs, multithread=<span class="hljs-literal">TRUE</span>)
errR &lt;- learnErrors(filtRs, multithread=<span class="hljs-literal">TRUE</span>)

<span class="hljs-comment"># make a pdf of the error output plot_errors.pdf</span>
pdf(file = paste(outpath,<span class="hljs-string">&quot;/plot_errors_F.pdf&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>), <span class="hljs-number">7</span>, <span class="hljs-number">7</span>)
plotErrors(errF, nominalQ=<span class="hljs-literal">TRUE</span>)
dev.off()
pdf(file = paste(outpath,<span class="hljs-string">&quot;/plot_errors_R.pdf&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>), <span class="hljs-number">7</span>, <span class="hljs-number">7</span>)
plotErrors(errR, nominalQ=<span class="hljs-literal">TRUE</span>)
dev.off()
</div></code></pre>
<h2 id="45-dereplicate-reads">4.5 Dereplicate reads</h2>
<p>The function <code>derepFastq()</code> will group all the identical reads</p>
<pre><code class="language-R"><div>derep_forward &lt;- derepFastq(filtFs, verbose=<span class="hljs-literal">TRUE</span>)
<span class="hljs-comment"># name the derep-class objects by the sample names</span>
<span class="hljs-built_in">names</span>(derep_forward) &lt;- sample.names

derep_reverse &lt;- derepFastq(filtRs, verbose=<span class="hljs-literal">TRUE</span>)
<span class="hljs-comment"># name the derep-class objects by the sample names</span>
<span class="hljs-built_in">names</span>(derep_reverse) &lt;- sample.names
</div></code></pre>
<h2 id="46-sample-inference">4.6 Sample Inference</h2>
<p>This is the main function of Dada2.</p>
<pre><code class="language-R"><div><span class="hljs-comment"># Forward - this clusters forward reads into ASV&#x27;s. later you will merge forward and reverse reads</span>
dadaFs &lt;- dada(derep_forward, err=errF, multithread=<span class="hljs-literal">TRUE</span>)
<span class="hljs-comment"># Reverse - this clusters reverse reads into ASV&#x27;s.</span>
dadaRs &lt;- dada(derep_reverse, err=errR, multithread=<span class="hljs-literal">TRUE</span>)

<span class="hljs-comment"># merged paired read data</span>
mergers &lt;- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=<span class="hljs-literal">TRUE</span>)
<span class="hljs-comment"># Inspect the merger data.frame from the first sample</span>
head(mergers[[<span class="hljs-number">1</span>]])

<span class="hljs-comment"># amplicon sequence variant table (ASV)</span>
seqtab &lt;- makeSequenceTable(mergers)
<span class="hljs-built_in">dim</span>(seqtab) <span class="hljs-comment"># check size of table</span>
<span class="hljs-comment"># Inspect distribution of sequence lengths</span>
table(nchar(getSequences(seqtab)))
</div></code></pre>
<h2 id="47-remove-chimeras">4.7 Remove chimeras</h2>
<p>We need to remove chimeras. A function in dada2 called <code>removeBimeraDenovo</code> will do this by removing reads that look like a hybrid of two others in the dataset. Chimeras made from more than two sequences merging are rare.</p>
<pre><code class="language-R"><div><span class="hljs-comment"># run removeBimeraDenovo</span>
seqtab.nochim &lt;- removeBimeraDenovo(seqtab, method=<span class="hljs-string">&quot;consensus&quot;</span>, multithread=<span class="hljs-literal">TRUE</span>, verbose=<span class="hljs-literal">TRUE</span>)

<span class="hljs-comment"># check size of data.frame</span>
<span class="hljs-built_in">dim</span>(seqtab.nochim)

<span class="hljs-comment"># this line will print a messgae uisng the outputs:</span>
print(paste(<span class="hljs-built_in">round</span>(<span class="hljs-built_in">sum</span>(seqtab.nochim)/<span class="hljs-built_in">sum</span>(seqtab),<span class="hljs-number">4</span>),<span class="hljs-string">&quot;% of ASV&#x27;s remain fter chimera removal: remaining ASV&#x27;s:&quot;</span>,<span class="hljs-built_in">dim</span>(seqtab.nochim[<span class="hljs-number">2</span>]),sep=<span class="hljs-string">&quot;&quot;</span>)) <span class="hljs-comment"># check proportion of reads are left after chimeras removed</span>

<span class="hljs-comment"># Save seqtab.nochim as an RDS file</span>
saveRDS(seqtab.nochim, paste(outpath,<span class="hljs-string">&quot;/seqtab_nochim.rds&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>))
</div></code></pre>
<h2 id="48-removal-of-negative-controls">4.8 Removal of negative controls</h2>
<p>If you have included negative controls in your experiment, you can remove the ASV's found in those control samples from the rest of the dataset. This is assuming that these ASV's have been introduced to the rest of your data through contamination. This method may be too simple if there are contaminants that are also present in the sample.</p>
<p>Here we will pretend that sample Ts16s is a negative control.</p>
<pre><code class="language-R"><div><span class="hljs-comment"># make a list of the negative control samples</span>
negative_controls&lt;-<span class="hljs-built_in">c</span>(<span class="hljs-string">&quot;T16s&quot;</span>)

<span class="hljs-comment"># save a new version of the ASV table `seqtab.nochim`</span>
seqtab.nochim.c&lt;-seqtab.nochim

<span class="hljs-comment"># this loop will loop through each negative control (f) in the list (negative controls) and remove all ASV&#x27;s found in that sample (to_remove) from the rest of the samples</span>
<span class="hljs-keyword">for</span>(f <span class="hljs-keyword">in</span> negative_controls){
	<span class="hljs-comment"># print which negative control is being used in the loop righ now</span>
  print(f)
	<span class="hljs-comment"># make a list of rows to drop so we can count how many</span>
  to_remove&lt;-<span class="hljs-built_in">dim</span>(seqtab.nochim.c[,which(seqtab.nochim.c[which(row.names(seqtab.nochim.c)==f),]&gt;<span class="hljs-number">0</span>)])[<span class="hljs-number">2</span>]
	<span class="hljs-comment"># keeps rows where negative control is zero</span>
  seqtab.nochim.c&lt;-seqtab.nochim.c[,which(seqtab.nochim.c[which(row.names(seqtab.nochim.c)==f),]&lt;=<span class="hljs-number">0</span>)]
	<span class="hljs-comment"># print a message to say how many ASV&#x27;s were dropped and kept</span>
  print(paste(<span class="hljs-built_in">dim</span>(seqtab.nochim.c)[<span class="hljs-number">2</span>],<span class="hljs-string">&quot; ASV&#x27;s remain after removing&quot;</span>, to_remove,<span class="hljs-string">&quot; ASV&#x27;s present in negative controls:&quot;</span>, f))
}

<span class="hljs-comment"># if we would like to keep this new ASV table, we can replace the old dataframe seqtab.nochim with the new one seqtab.nochim.c and proceed:</span>
<span class="hljs-comment"># seqtab.nochim&lt;-seqtab.nochim.c</span>
<span class="hljs-comment"># rm(seqtab.nochim.c)</span>
<span class="hljs-comment"># we won&#x27;t do this here because we didn&#x27;t use a true negative control</span>

<span class="hljs-comment"># save result</span>
write.table(seqtab.nochim, file = paste(outpath,<span class="hljs-string">&quot;/asv_table.dada2.tsv&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>),sep=<span class="hljs-string">&quot;\t&quot;</span>,row.names=<span class="hljs-built_in">T</span>)

<span class="hljs-comment"># save csv for MicrobiomeAnalyst</span>
<span class="hljs-comment"># transpose</span>
seqtab.nochim.c.df &lt;- t(seqtab.nochim.c)
<span class="hljs-comment"># make a column for the samples names called &quot;#NAMES&quot;</span>
seqtab.nochim.c.df &lt;- cbind(row.names(seqtab.nochim.c.df),seqtab.nochim.c.df)

<span class="hljs-comment"># set first column name</span>
colnames(seqtab.nochim.c.df)[<span class="hljs-number">1</span>]&lt;-<span class="hljs-string">&quot;#NAMES&quot;</span>

<span class="hljs-comment"># write output</span>
write.csv(seqtab.nochim.c.df, file = paste(outpath,<span class="hljs-string">&quot;/asv_table.dada2.ma.csv&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>),row.names=<span class="hljs-built_in">F</span>)
</div></code></pre>
<h2 id="49-track-data-across-each-step">4.9 Track data across each step</h2>
<p>Let's build a table that shows how many reads/ASV's are left after each step.</p>
<p>You will need to remove samples from &quot;out&quot;, Ts29_1</p>
<blockquote>
<p>For example to remove sample &quot;sample_name1 and &quot;sample_name2&quot;:
<code>out &lt;- out[which(row.names(out)!=&quot;sample_name1&quot; &amp; row.names(out)!=&quot;sample_name2&quot;),]</code></p>
</blockquote>
<pre><code class="language-R"><div><span class="hljs-comment"># set getN function - this function will count for us</span>
getN &lt;- <span class="hljs-keyword">function</span>(x) <span class="hljs-built_in">sum</span>(getUniques(x))

<span class="hljs-comment"># If you removed samples above you need to remove them again here:</span>
out &lt;- out[which(row.names(out)!=<span class="hljs-string">&quot;Ts29_1&quot;</span> &amp; row.names(out)!=<span class="hljs-string">&quot;sample_name2&quot;</span> ),]

<span class="hljs-comment"># Combine filtering, dada2, and chimera removal data into one dataframe</span>
<span class="hljs-comment"># If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)</span>
track &lt;- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim), rowSums(seqtab.nochim.c))

<span class="hljs-comment"># set some column names</span>
colnames(track) &lt;- <span class="hljs-built_in">c</span>(<span class="hljs-string">&quot;no_reads_input&quot;</span>, <span class="hljs-string">&quot;no_reads_after_filter&quot;</span>, <span class="hljs-string">&quot;no_ASV_denoisedF&quot;</span>, <span class="hljs-string">&quot;no_ASV_denoisedR&quot;</span>, <span class="hljs-string">&quot;no_ASV_merged&quot;</span>, <span class="hljs-string">&quot;no_ASV_nonchim&quot;</span>,<span class="hljs-string">&quot;no_ASV_neg_cntrl_removed&quot;</span>)
rownames(track) &lt;- sample.names

<span class="hljs-comment"># check output</span>
head(track) <span class="hljs-comment"># head shows the first 10 entries</span>

<span class="hljs-comment"># write to file:</span>
write.table(track, paste(outpath,<span class="hljs-string">&quot;/track_reads.tsv&quot;</span>, sep=<span class="hljs-string">&quot;&quot;</span>), sep=<span class="hljs-string">&quot;\t&quot;</span>)
</div></code></pre>
<blockquote>
<p>We can view this table in order of reads remain after negative controls are removed as follows:
<code>track[order(track[,7], decreasing=T),]</code></p>
</blockquote>
<h2 id="410-taxonomy">4.10 Taxonomy</h2>
<p>Next, we will assign taxa to each ASV we have detected. We will compare our ASV's to he reference database SILVA.</p>
<p>Functions for assigning taxonomy are <code>taxa_with_bootstraps()</code> and <code>taxa()</code>. Here we use <code>taxa_with_bootstraps()</code>.</p>
<pre><code class="language-R"><div><span class="hljs-comment"># dataset - SILVA:</span>
taxa_with_bootstraps &lt;- assignTaxonomy(seqtab.nochim, <span class="hljs-string">&quot;~/metabarcoding_ws/db/silva_nr99_v138.1_train_set.fa.gz&quot;</span>, outputBootstraps=<span class="hljs-literal">TRUE</span>, multithread=<span class="hljs-literal">TRUE</span>)

<span class="hljs-comment"># we can save the taxa dataframe without the bootstraps:</span>
taxa_all &lt;- taxa_with_bootstraps$tax

<span class="hljs-comment"># optionally add species:</span>
taxa_species &lt;- addSpecies(taxa_with_bootstraps$tax, <span class="hljs-string">&quot;~/metabarcoding_ws/db/silva_species_assignment_v138.1.fa.gz&quot;</span>)

<span class="hljs-comment"># write to files</span>
write.table(taxa_with_bootstraps, file = paste(outpath,<span class="hljs-string">&quot;/taxa_with_bootstraps.tsv&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>),sep=<span class="hljs-string">&quot;\t&quot;</span>,row.names=<span class="hljs-built_in">T</span>)
write.table(taxa_all, file = paste(outpath,<span class="hljs-string">&quot;/taxa.tsv&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>),sep=<span class="hljs-string">&quot;\t&quot;</span>,row.names=<span class="hljs-built_in">T</span>)
write.table(taxa_species, file = paste(outpath,<span class="hljs-string">&quot;/taxa_species.tsv&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>),sep=<span class="hljs-string">&quot;\t&quot;</span>,row.names=<span class="hljs-built_in">T</span>)

write.csv(taxa_with_bootstraps, file = paste(outpath,<span class="hljs-string">&quot;/taxa_with_bootstraps.csv&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>),row.names=<span class="hljs-built_in">T</span>)
write.csv(taxa_all, file = paste(outpath,<span class="hljs-string">&quot;/taxa.csv&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>),row.names=<span class="hljs-built_in">T</span>)
write.csv(taxa_species, file = paste(outpath,<span class="hljs-string">&quot;/taxa_species.csv&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>),row.names=<span class="hljs-built_in">T</span>)


<span class="hljs-comment"># write for MicrobiomeAnalyst</span>
<span class="hljs-built_in">names</span>(axa_all&lt;-
write.csv(taxa_all, file = paste(outpath,<span class="hljs-string">&quot;/taxa.ma.csv&quot;</span>,sep=<span class="hljs-string">&quot;&quot;</span>),row.names=<span class="hljs-built_in">T</span>)

</div></code></pre>
<p>Now you have completed the Dada2 aspect of the analysis!</p>
<blockquote>
<p>You should now  have been left with output files in <code>/home/$USER/metabarcoding_ws/outputs/dada2/</code></p>
<p><strong>Most important:</strong>
ASV file: <code>asv_table.dada2.tsv</code>
Taxonomy files: <code>taxa.tsv</code>, <code>taxa_with_bootstraps.tsv</code>, and <code>taxa_species.tsv</code></p>
<p><strong>Additional:</strong>
Filtered reads directory: <code>filtered/</code>
Read quality pdfs (pre filtering): <code>plot_read_quality_F.pdf</code>, <code>plot_read_quality_R.pdf</code>
Read quality pdfs (post filtering): <code>plot_read_quality_filtered_F.pdf</code>, <code>plot_read_quality_filtered_R.pdf</code>
R formatted ASV table pre contaminamt removal: <code>seqtab_nochim.rds</code>
Track reads file: <code>track_reads.tsv</code>
Saved out file (can be discarded now): <code>out_save_F1.tsv</code>
Error rate plot pdfs: <code>plot_errors_F.pdf</code>, <code>plot_errors_R.pdf</code></p>
</blockquote>
<h2 id="5-alternatively-run-dadaist2">5. Alternatively run <a href="">dadaist2</a></h2>
<h4 id="in-the-command-line">In the command-line:</h4>
<pre><code><code><div>msub -I -q S30 -l procs=8,walltime=1:00:00 # change depending on what you think you need.

in_dir=~/metabarcoding_ws/data/16S/fastq
out_dir=~/metabarcoding_ws/outputs/dadaist2
database=~/metabarcoding_ws/db/silva_nr99_v138.1_train_set.fa.gz
meta=~/metabarcoding_ws/outputs/dadaist2/metadatafile.csv # make one using dadaist2-metadata below
# activate conda environment
conda activate dadaist2

# optionally move into project directory
cd $in_dir
# make output directory
mkdir $out_dir

# make a metadata file if one has not already been made
dadaist2-metadata -i $in_dir -o $meta -1 _1 -2 _2

# main command - check parameters
# note - if primers not supplied switch on fastp trimming. It will skip trimming if primers are not supplied and cutadapt trimming is selected.
dadaist2 \
-input-directory $in_dir  \
-output-directory $out_dir \
-database  $database \
-metadata $meta \
-threads 8 \
-trunc-len-1 200 \
-trunc-len-2 160 \
-s1 0 \
-s2 0 \
-min-qual 28 \
-maxee1 2 \
-maxee2 2 \
-save-rds \
-1 _1 \
-2 _2 \
-verbose

# export to get MetagenomeAnalyist compatable files
dadaist2-exporter -i $out_dir
# make a multiqc report
dadaist2-mqc-report  -i $out_dir  -o $out_dir/multiqc
# find alpha diversities
dadaist2-normalize  -i $out_dir/feature-table.tsv -o $out_dir/normalise
# make a phyloseq object to import to R
dadaist2-phyloseqMake -i $out_dir
# use phyloseq object to automatically produce figures
dadaist2-taxaplot [options] -i phyloseq.rds -o $out_dir/plots
</div></code></code></pre>
<h1 id="6-visualisation-and-statistics">6. Visualisation and statistics:</h1>
<h2 id="61-microbiomeanalyst">6.1 <a href="https://www.microbiomeanalyst.ca/">MicrobiomeAnalyst</a></h2>
<p>To submit the files in the Dadaist MicrobiomeAnalyst directory you will first need to edit some of them.</p>
<p>The table.txt file needs to have file extensions removed from the sample names (use find and replace).</p>
<p>The taxa file doesn't contain enough delimiters. open in excel and save as csv file.</p>
<p>Download metadata file:</p>
<pre><code><code><div>wget https://raw.githubusercontent.com/nmc97/metabarcoding_at_cefas_tutorial/main/scripts/workshop_1/data/metadata.csv
</div></code></code></pre>
<h2 id="62-phyloseq">6.2 Phyloseq</h2>
<p>For this you need R. It would be best to set up r studio on your computer and run this there.</p>
<p>You will need to create a Phyloseq object from the outputs of Dada2 or Dadaist2</p>
<p>For example: Dadaist two code for plots: <a href="https://quadram-institute-bioscience.github.io/dadaist2/notes/plot.html">https://quadram-institute-bioscience.github.io/dadaist2/notes/plot.html</a></p>
<h2 id="63-microbiome-r-package">6.3 Microbiome R package</h2>
<p>For this you need R. It would be best to set up r studio on your computer and run this there.</p>
<p>You will need to create a Phyloseq object from the outputs of Dada2 or Dadaist2</p>
<p><a href="https://microbiome.github.io/tutorials/">https://microbiome.github.io/tutorials/</a></p>
<h3 id="623-installation">6.2.3 Installation</h3>
<pre><code><code><div>library(BiocManager)
BiocManager::install(&quot;microbiome&quot;)
</div></code></code></pre>
<p>Or</p>
<pre><code><code><div>library(devtools) # Load the devtools package
install_github(&quot;microbiome/microbiome&quot;) # Install the package
</div></code></code></pre>

        <script async src="https://cdn.jsdelivr.net/npm/katex-copytex@latest/dist/katex-copytex.min.js"></script>
        
    </body>
    </html>