---
title: "microbiome_r_package_plotting.md"
author: "Nicola Coyle"
date: "17/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Microbiom R package plotting

You may need to modify to apply to your data

Prerequisites :

- microbiome R package
- microbiomeultilities
- dplyr
- hrbrthemes

**Install using code below:**

```
# Install microbiome
library(BiocManager)
BiocManager::install("microbiome")

# Install microbiomeutilities
install.packages("devtools")
devtools::install_github("microsud/microbiomeutilities")

# Install dplyr
install.packages("dplyr")

# Intall hrbrthemes
install.packages("hrbrthemes")
install.packages("gcookbook")
install.packages("tidyverse")
```


```{r echo=FALSE}

# load libraries
library("microbiome")
library("microbiomeutilities")
library("dplyr")
library("knitr")
library("hrbrthemes")
library("gcookbook")
library("tidyverse")

```


# Importing files

Examples of importing files:

```
# Import output CSV files generated by write_phyloseq
datos1 <- read_phyloseq(otu.file, taxonomy.file, metadata.file, type = "simple")

# Import mother .shared and .taxonomy and metadata files
datos2 <- read_phyloseq(otu.file, taxonomy.file, metadata.file, type = "mothur")

# Import BIOM files
datos3 <- read_phyloseq(otu.file, taxonomy.file, metadata.file, type = "biom")

# Alternatively, you can read your data in R (read.table, read.csv or other standard functions) and convert into phyloseq format.
# http://joey711.github.io/phyloseq/import-data
```



```{r , echo=FALSE}


#================================================#
# Importing standard formats (CSV, Mothur, BIOM)
#================================================#

# using MircobiomeAnalyist ready files:

#otu.file <- "C:/Users/NC07/OneDrive - CEFAS/mobaxterm_home/MicrobiomeAnalyst/table.csv"
#taxonomy.file <- "C:/Users/NC07/OneDrive - CEFAS/mobaxterm_home/MicrobiomeAnalyst/taxonomy.csv"
#metadata.file <- "C:/Users/NC07/OneDrive - CEFAS/mobaxterm_home/MicrobiomeAnalyst/metadata.csv"
#data_type<-"simple"

# otu.file <- "C:/Users/NC07/OneDrive - CEFAS/Projects/Metabarcoding/Gene_pools/dadaist2/MicrobiomeAnalyst/MicrobiomeAnalyst/MicrobiomeAnalyst/table.csv"
# taxonomy.file <- "C:/Users/NC07/OneDrive - CEFAS/Projects/Metabarcoding/Gene_pools/dadaist2/MicrobiomeAnalyst/MicrobiomeAnalyst/MicrobiomeAnalyst/taxonomy_saved.csv"
# metadata.file <- "C:/Users/NC07/OneDrive - CEFAS/Projects/Metabarcoding/Gene_pools/dadaist2/MicrobiomeAnalyst/MicrobiomeAnalyst/MicrobiomeAnalyst/metadata_simgroup.csv"
# data_type<-"simple"

otu.file <- "C:/Users/NC07/OneDrive - CEFAS/Projects/Metabarcoding/Gene_pools/test_microbiome_package/MicrobiomeAnalyst/asv_table.MA.dada2.csv"
taxonomy.file <- "C:/Users/NC07/OneDrive - CEFAS/Projects/Metabarcoding/Gene_pools/test_microbiome_package/MicrobiomeAnalyst/taxa.species.MA.csv"
metadata.file <- "C:/Users/NC07/OneDrive - CEFAS/Projects/Metabarcoding/Gene_pools/test_microbiome_package/MicrobiomeAnalyst/metadata.csv"
data_type<-"simple"


# Import output CSV files generated by write_phyloseq

print(paste("Reading files:", sep=" "))
print(paste("OTU table -", otu.file, sep=" "))
print(paste("Taxonomy table -", taxonomy.file, sep=" "))
print(paste("Metadata file -", metadata.file, sep=" "))

datos_phyloseq <- read_phyloseq(otu.file, taxonomy.file, metadata.file, type = data_type)

sample_names(datos_phyloseq)

```

# Explore data

Your data Looks like:

```{r , echo=FALSE, warning=FALSE, message = FALSE, error= FALSE}


#================================================#
# initial look
#================================================#

summary<-summarize_phyloseq(datos)
kable(t(as.data.frame(summarize_phyloseq(datos)))[,1], row.names = FALSE)


# metadata extract
meta <- sample_data(datos)

# taxtable extract
taxonomy <- tax_table(datos)

# Absolute abundances
otu.absolute <- abundances(datos)

# Relative abundances
otu.relative <- abundances(datos, "compositional")

#Total read counts:
reads_sample <- readcount(datos)
# check for first 5 samples
reads_sample[1:5]

# Add read per sample to phyloseq object metadata.
sample_data(datos)$reads_sample <- reads_sample

## reads_sample is add to the last column in sample_data of datos object.
head(meta(datos)[,c("sample", "reads_sample")])

# melt for plotting
df <- psmelt(datos)
kable(head(df))

# sample sums function - how many reads per sample
s <- sample_sums(datos)

# look for specific taxa
head(abundances(datos)["c__Parcubacteria",])
abundances(datos)[1:10,1:10]

# subset
datos.subset <- subset_samples(datos, nationality == "AFR")

# datos
sample_variables(datos)

p <- plot_frequencies(sample_data(datos), "Files", "reads_sample")
print(p)

tab <-microbiome::alpha(datos, index = "all")
kable(head(tab))
tab <- dominance(datos, index = "all")


# # get values for first variable
# head(get_variable(datos, sample_variables(datos)[1]))
#
# # Calculate diversity for samples
# div <- microbiome::alpha(datos, index = "shannon")
#
# # Assign the estimated diversity to sample metadata
# sample_data(datos)$diversity <- div
#
# #Taxa operations
# #Number of taxa
# n <- ntaxa(datos)
#
# #Most abundant taxa
# topx <- top_taxa(datos, n = 10)
#
# #Names
# ranks <- rank_names(datos)  # Taxonomic levels
# taxa  <- taxa(datos)        # Taxa names at the analysed level
#
# #Subset taxa
# datos.bac <- subset_taxa(datos, Phylum == "Bacteroidetes")
# #Prune (select) taxa:
#
# # List of Genera in the Bacteroideted Phylum
# taxa <- map_levels(NULL, "Phylum", "Genus", datos)$Bacteroidetes
#
# # With given taxon names
# ex2 <- prune_taxa(taxa, datos)
#
# # Taxa with positive sum across samples
# ex3 <- prune_taxa(taxa_sums(datos) > 0, datos)
#
# #Filter by user-specified function values (here variance):
# f <- filter_taxa(datos, function(x) var(x) > 1e-05, TRUE)
#
# #List unique phylum-level groups:
# head(get_taxa_unique(datos, "Phylum"))
#
# ## [1] "Actinobacteria"  "Firmicutes"      "Proteobacteria"  "Verrucomicrobia" "Bacteroidetes"   "Spirochaetes"
# #Pick the taxa abundances for a given sample:
# samplename <- sample_names(datos)[[1]]
#
# # Pick abundances for a particular taxon
# tax.abundances <- abundances(datos)[, samplename]
#

```



```{r , echo=FALSE}

ps1.rel <- microbiome::transform(datos, "compositional")

ps1.fam.rel <-aggregate_rare(ps1.rel, level = "Family", detection = 0.005, prevalence = 0.5)

p <- plot_composition(ps1.fam.rel,
                      average_by = "City") +
  guides(fill = guide_legend(ncol = 1)) +
  labs(x = "Samples",
       y = "Relative abundance",
       title = "Relative abundance data",
       subtitle = "Subtitle",
       caption = "Caption text.")
print(p + scale_fill_brewer("Family", palette = "Paired") + theme_bw())


p.famrel <- plot_composition(ps1.fam.rel,
                             sample.sort = NULL,
                             otu.sort = NULL,
                             x.label = "City",
                             plot.type = "barplot",
                             verbose = FALSE) +
  theme_minimal() +
  guides(fill = guide_legend(ncol = 1)) +
  labs(x = "Animal secretion samples",
       y = "Relative abundance",
       title = "Relative abundance data",
       subtitle = "Subtitle can be added here",
       caption = "Caption text can be added here.") +
  scale_fill_brewer("Family", palette = "Paired") +

  #Removes sample names and ticks
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  #Adjusts size of subtitle, caption, legend text and legend title
  theme(plot.subtitle=element_text(size=8),
        plot.caption=element_text(size=8),
        legend.text=element_text(size=8),
        legend.title =element_text(size=10))

print(p.famrel)

```



Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
